File: test/tests/buffer.cc
#include <fixtures/buffer.h>

#include <nobind.h>

Nobind::Typemap::Buffer nobind_get_buffer() {
  Nobind::Typemap::Buffer buf{nullptr, 16};
  get_buffer(buf.first, buf.second, 0x17);
  return buf;
}
void nobind_put_buffer(Nobind::Typemap::Buffer buf) {
  put_buffer(buf.first, buf.second, 0x17);
}

NOBIND_MODULE(buffer, m) {
  m.def<&nobind_get_buffer>("get_buffer")
  .def<&nobind_put_buffer>("put_buffer");
}

File: test/tests/buffer.js
const { assert } = require('chai');

describe('get', () => {
  it('nominal', () => {
    const buf = dll.get_buffer();
    assert.instanceOf(buf, Buffer);
    assert.lengthOf(buf, 16);
    const typed = new Uint32Array(buf.buffer);
    assert.deepStrictEqual(typed, new Uint32Array([0x17, 0x17, 0x17, 0x17]));
  });
});

describe('put', () => {
  it('nominal', () => {
    const buf = Buffer.from(new Uint32Array([0x17, 0x17, 0x17, 0x17]).buffer);
    dll.put_buffer(buf);
  });
  it('exception', () => {
    assert.throws(() => {
      const buf = Buffer.from(new Uint32Array([0x17, 0x17, 0x11, 0x17]).buffer);
      dll.put_buffer(buf);
    }, /Invalid value/);
  });
});
